awswrangler/cleanrooms/_read.py

```diff
@@ -0,0 +1,128 @@
+        `:name`. Note that for varchar columns and similar, you must surround the value in single quotes
+    chunksize : Union[int, bool], optional
+        If passed, the data is split into an iterable of DataFrames (Memory friendly).
+        If `True` an iterable of DataFrames is returned without guarantee of chunksize.
+        If an `INTEGER` is passed, an iterable of DataFrames is returned with maximum rows
+        equal to the received INTEGER
+    use_threads : Union[bool, int], optional
+        True to enable concurrent requests, False to disable multiple threads.
+        If enabled os.cpu_count() is used as the maximum number of threads.
+        If integer is provided, specified number is used
+    boto3_session : boto3.Session, optional
+        Boto3 Session. If None, the default boto3 session is used
+    pyarrow_additional_kwargs : Optional[Dict[str, Any]]
+        Forwarded to `to_pandas` method converting from PyArrow tables to Pandas DataFrame.
+        Valid values include "split_blocks", "self_destruct", "ignore_metadata".
+        e.g. pyarrow_additional_kwargs={'split_blocks': True}
+
+    Returns
+    -------
+    Union[Iterator[pd.DataFrame], pd.DataFrame]
+        Pandas DataFrame or Generator of Pandas DataFrames if chunksize is provided.
+
+    Examples
+    --------
+    >>> import awswrangler as wr
+    >>> df = wr.cleanrooms.read_sql_query(
+    >>>     sql='SELECT DISTINCT...',
+    >>>     membership_id='membership-id',
+    >>>     output_bucket='output-bucket',
+    >>>     output_prefix='output-prefix',
+    >>> )
+    """
+    client_cleanrooms = _utils.client(service_name="cleanrooms", session=boto3_session)
+
+    query_id: str = client_cleanrooms.start_protected_query(
+        type="SQL",
+        membershipIdentifier=membership_id,
+        sqlParameters={"queryString": _process_sql_params(sql, params, engine_type="partiql")},
+        resultConfiguration={
+            "outputConfiguration": {
+                "s3": {
+                    "bucket": output_bucket,
+                    "keyPrefix": output_prefix,
+                    "resultFormat": "PARQUET",
+                }
+            }
+        },
+    )["protectedQuery"]["id"]
+
+    _logger.debug("query_id: %s", query_id)
+    path: str = wait_query(membership_id=membership_id, query_id=query_id)["protectedQuery"]["result"]["output"]["s3"][
+        "location"
+    ]
+
+    _logger.debug("path: %s", path)
+    chunked: Union[bool, int] = False if chunksize is None else chunksize
+    ret = s3.read_parquet(
+        path=path,
+        use_threads=use_threads,
+        chunked=chunked,
+        boto3_session=boto3_session,
+        pyarrow_additional_kwargs=pyarrow_additional_kwargs,
+    )
+
+    _logger.debug("type(ret): %s", type(ret))
+    kwargs: Dict[str, Any] = {
+        "path": path,
+        "use_threads": use_threads,
+        "boto3_session": boto3_session,
+    }
+    if chunked is False:
```

Comment:
What do you think about passing `chunked` to `_delete_after_iterate` and let it return/yield and handle the cleanup in both scenarios? Less LOC and only one cleanup

Reply:
AFAIK you cannot have a yield and return statement defined in the same function. So I would to split the two somehow. The other reason is that if you look carefully for the return we delete and then return (as I cannot delete after returning), while for the yield we yield first then delete (as we cannot delete until everything was yielded). But please share the logic if you have a way around this

Reply:
I suppose you can put the yields into an enclosed function and then the logic would be iterate (if needed), clean, return